### PR TITLE
lightweight isl::id

### PR DESCRIPTION
**Pull request by ftynse**
All the lifetime management is done by the C side.  Therefore, the deleter must
be implicitly convertible to a C function pointer (no capturing lambdas or
binders).  By default, the deleter does nothing to be consistent with the C
interface.  We may additionally provide a template constructor variant that
provides a lambda with "delete static_cast<T*>(v)" inside.

### CODE REVIEW

**Comment by ftynse** [comment]
Most of functions in `isl_id` class are boilerplate and can be generated.  We will need to inject the templates, `has_name` and comparison operators.

**Comment by ftynse** [comment]
This is what we discussed this morning with @nicolasvasilache and @albertcohen.

File: interface/isl.h.top [hunk 1]
```diff
@@ -87,6 +88,154 @@ public:
...
+
+__isl_give isl_id *id::release() {
+  isl_id *tmp = ptr;
+  ptr = nullptr;
+  return tmp;
+}
+
+__isl_keep isl_id *id::keep() const {
+  return get();
+}
```
Comment (tobig) [inline interface/isl.h.top]: You should really drop the redundant interface definitions which I only added to my generator for Polly as we used a different naming style beforehand.

File: interface/isl.h.top [hunk 2]
```diff
@@ -87,6 +88,154 @@ public:
...
+
+id id::set_free_user(void (*deleter)(void *)) const {
+  auto res = isl_id_set_free_user(copy(), deleter);
+  return manage(res);
+}
+
+bool id::operator==(const id &obj) const {
+  return ptr == obj.ptr;
+}
```
Comment (tobig) [inline interface/isl.h.top]: If you want to use this in sets, we likely want an implementation of operator< or less(A,B).

https://stackoverflow.com/a/1114862

File: interface/isl.h.top [hunk 3]
```diff
@@ -87,6 +88,154 @@ public:
...
+
+bool id::has_name() const {
+  return isl_id_get_name(get()) != nullptr;
+}
+
+id id::set_free_user(void (*deleter)(void *)) const {
+  auto res = isl_id_set_free_user(copy(), deleter);
+  return manage(res);
+}
```
Comment (tobig) [inline interface/isl.h.top]: Is this auto-generated or did you write this by hand?

Comment (ftynse) [inline interface/isl.h.top]: This particular function is hand-written.  Most others are monkey-generated, i.e. I patched the generator quick-and-dirty to produce the boilerplate I needed.

File: interface/isl.h.top (hunk 2 above)
Comment (ftynse) [inline interface/isl.h.top]: `std::set` may be instantiated with a custom `Compare` type.
I don't think exposing less-than comparison of IDs makes sense in general.  One could assume it is lexicographical comparison of their names, for example.

I would rather consider something like
```c++
template <typename T>
struct isl_object_set_compare {
  bool operator() (T left, T right) {
    return left.get() == right.get();
  }
};
```
which could be used as
```c++
std::set<isl::set, isl::isl_object_set_compare<isl::set>>
```

On the other hand, I would specialize `std::hash` for unordered stuff.

File: interface/isl.h.top (hunk 3 above)
Comment (tobig) [inline interface/isl.h.top]: You certainly achieved the "quick-and-dirty". ;-) It is likely not very difficult to auto-generate even this function.

File: interface/isl.h.top (hunk 2 above)
Comment (tobig) [inline interface/isl.h.top]: These are also interesting ideas. I personally don't feel very strong about this. If we want to get a minimal patch for this to be upstreamed, it is likely a good idea to leave comparisons out for for now.

Comment (ftynse) [inline interface/isl.h.top]: This is now https://github.com/PollyLabs/isl/issues/13

**Pull request merged**